File: dptemple/src/main/java/nlu/fit/dptemple/service/NotificationService.java
package nlu.fit.dptemple.service;

import nlu.fit.dptemple.entity.Notification;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.Optional;

public interface NotificationService {

    Notification create(Notification notification);

    Notification update(String id, Notification notification);

    void delete(String id, String deletedById);

    Optional<Notification> findById(String id);

    List<Notification> findHomepageNotifications();

    List<Notification> findHomepageNotificationsLimited();

    List<Notification> findFeaturedNotifications();

    List<Notification> findFeaturedNotificationsLimited();

    Page<Notification> findAllPublished(Pageable pageable);

    Page<Notification> findAllDeleted(Pageable pageable);
}

File: dptemple_ui/src/services/notificationService.ts
import { axiosInstance } from '@/lib/axios'
import type {
  Notification,
  PageNotification,
  Pageable
} from '@/types'

export const notificationService = {
  // Get all published notifications with pagination
  findAllPublished: (pageable: Pageable) =>
    axiosInstance
      .get<PageNotification>('/notifications', { params: pageable })
      .then((r) => r.data),

  // Create a new notification
  create: (data: Omit<Notification, 'id' | 'createdAt' | 'updatedAt'>) =>
    axiosInstance
      .post<Notification>('/notifications', data)
      .then((r) => r.data),

  // Get notification by ID
  findById: (id: string) =>
    axiosInstance
      .get<Notification>(`/notifications/${id}`)
      .then((r) => r.data),

  // Update notification
  update: (id: string, data: Partial<Notification>) =>
    axiosInstance
      .put<Notification>(`/notifications/${id}`, data)
      .then((r) => r.data),

  // Soft delete notification
  delete: (id: string, deletedById: string) =>
    axiosInstance
      .delete<void>(`/notifications/${id}`, { params: { arg1: deletedById } })
      .then((r) => r.data),

  // Get homepage notifications
  findHomepage: () =>
    axiosInstance
      .get<Notification[]>('/notifications/homepage')
      .then((r) => r.data),

  // Get featured notifications
  findFeatured: () =>
    axiosInstance
      .get<Notification[]>('/notifications/featured')
      .then((r) => r.data),

  // Get all deleted notifications
  findAllDeleted: (pageable: Pageable) =>
    axiosInstance
      .get<PageNotification>('/notifications/deleted', { params: pageable })
      .then((r) => r.data),
}